### PR TITLE
ThreadSanitizer support for POWER

### PR DESCRIPTION
**Pull request by ghost**
This adds TSan support to Linux/power.

This has been tested using `gcc version 11.4.0 (Ubuntu 11.4.0-1ubuntu1~22.04) ` and `clang version 18.0.0git (https://github.com/llvm/llvm-project.git d0f5039e5db5c2b9222f78e7242401061ab259dc)`.

The current TSan runtime is broken on POWER and will cause all uses of `fstat` to always return `EINVAL` when targeting glibc. Other libcs (i.e. musl) should be safe. This has been reported [here](https://github.com/llvm/llvm-project/issues/75346) and a trivial fix is available [here](https://github.com/llvm/llvm-project/pull/75578).

Using a fixed TSan runtime, all tests pass with clang; all tests but one pass with gcc, the failing test is `tests/tsan/perform.ml` where the traceback shows `caml_resume` with a huge offset instead of `caml_runstack` with a small offset; in other words, the address is correct and always within `caml_runstack`, but for some reason libunwind misbehaves here, this is probably not worth spending too much time investigating.

### CODE REVIEW

**Comment by dra27** [comment]
(added to 5.2 milestone for discussion later)

File: runtime/power.S [hunk 1]
```diff
@@ -321,6 +440,18 @@ FUNCTION caml_c_call
...
+        TSAN_SETUP_C_CALL 16
+    /* Save return value registers. Since the called function could be anything,
+       it may have returned its result (if any) either in r3 or f1. */
+        std     3,  (RESERVED_STACK + 0)(SP)
+        stfd    1,  (RESERVED_STACK + 8)(SP)
+        li      3, 0
+        Far_call(__tsan_func_exit)
+        lfd     1,  (RESERVED_STACK + 8)(SP)
+        ld      3,  (RESERVED_STACK + 0)(SP)
+        TSAN_CLEANUP_AFTER_C_CALL 16
```
Comment (tmcgilchrist) [inline runtime/power.S:453]: I was expecting a TSAN_EXIT_FUNCTION macro to be used here like arm64.S and amd64.S. It makes it a little clearer when comparing implementations on various architectures.

File: runtime/power.S [hunk 2]
```diff
@@ -702,6 +956,27 @@ FUNCTION caml_perform
...
+        mr      5, 10   /* arg3: performer stack */
+        SWITCH_OCAML_TO_C
+        TSAN_SETUP_C_CALL 0
+        Far_call(caml_tsan_entry_on_resume)
+        /* invoke __tsan_func_entry without shuffling stacks */
+        ld      3, LR_SAVE(SP)
+        Far_call(__tsan_func_entry)
+        TSAN_CLEANUP_AFTER_C_CALL 0
+        SWITCH_C_TO_OCAML
+        TSAN_RESTORE_CALLER_REGS
```
Comment (tmcgilchrist) [inline runtime/power.S:978]: I also expected to see a `TSAN_C_CALL` macro in places like this similar to other backends. Can you comment on why this is different or if it does not make sense for Power?

**Review by tmcgilchrist — COMMENTED** [review]
Running on gcc-11.4/Ubuntu 22.04/ppc64le I get these failures:
```
    tests/tsan/'exn_reraise.ml' with line 10 (native) 
    tests/tsan/'raise_through_handler.ml' with line 10 (native) 
    tests/tsan/'unhandled.ml' with line 10 (native) 
```
That is different to `tests/tsan/perform.ml` that you reported was failing @dustanddreams 
Minor comments about the use of macros, but the code looks reasonable to me.

File: runtime/power.S (hunk 1 above)
Comment (ghost) [reply runtime/power.S:453]: I did not define a `TSAN_EXIT_FUNCTION` on this platform. The various calling points all have slightly different stack frames, so the best we can do is `TSAN_SETUP_C_CALL` + frame setup + `Far_call` + frame undoing + `TSAN_CLEANUP_AFTER_C_CALL`.

File: runtime/power.S (hunk 2 above)
Comment (ghost) [reply runtime/power.S:978]: The 32 bytes from `0(sp)` onwards are reserved by the ABI. So any register which needs to be saved, has to be saved from `32(sp)` onwards, which in turn requires different values of stack adjustment depending on the need. This is why the `TSAN_SETUP_C_CALL` and `TSAN_CLEANUP_AFTER_C_CALL` macros take an argument, which is the amount of data needed, and you can't have a simple `TSAN_C_CALL` macro.

**Comment by ghost** [comment]
> Running on gcc-11.4/Ubuntu 22.04/ppc64le I get these failures:
> 
> ```
>     tests/tsan/'exn_reraise.ml' with line 10 (native) 
>     tests/tsan/'raise_through_handler.ml' with line 10 (native) 
>     tests/tsan/'unhandled.ml' with line 10 (native) 
> ```

These failures are caused by differences in backtraces, where libunwind returns wrong symbols. Not sure there is much which can be done here except wait for a toolchain update, especially with clang producing correct results.

**Review by tmcgilchrist — APPROVED** [review]

File: runtime/power.S (hunk 2 above)
Comment (tmcgilchrist) [reply runtime/power.S:978]: Makes total sense, thanks for the explanation.

**Comment by shindere** [comment]
Approving and merging on behalf of @tmcgilchrist

I guess this should be backported to 5.2?

**Review by shindere — APPROVED** [review]

**Pull request merged**